# Numpy

## Numpy의 소개

NumPy(Numerical Python)는 파이썬에서 과학적 계산을 위한 핵심 라이브러리이다. NumPy는 다차원 배열 객체와 배열과 함께 작동하는 도구들을 제공한다. 하지만 NumPy 자체로는 고수준의 데이터 분석 기능을 제공하지 않기 때문에 NumPy 배열과 배열 기반 컴퓨팅의 이해를 통해 pandas와 같은 도구를 좀 더 효율적으로 사용하는 것이 필요하다.

### ndarray 생성
array 함수를 사용해서 배열을 생성하기

In [1]:
import numpy as np

## ndarray 생성
arr = np.array([1,2,3,4])
print(arr)
## list와 다른 numpy 배열이다

[1 2 3 4]


In [2]:
## 0으로 초기화된 구성요소가 3개인 배열이 3개가 생성된다
np.zeros((3,3))

array([[0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.]])

In [3]:
## 1로 초기화된 구성요소가 2개인 배열 2개가 생성된다
np.ones((2,2))

array([[1., 1.],
       [1., 1.]])

In [4]:
## 초기화되지 않은 구성요소가 4개인 배열 4개가 생성된다
np.empty((4,4))

array([[6.23042070e-307, 4.67296746e-307, 1.69121096e-306,
        9.34613185e-307],
       [7.56587585e-307, 1.37961302e-306, 1.05699242e-307,
        8.01097889e-307],
       [1.78020169e-306, 7.56601165e-307, 1.02359984e-306,
        1.33510679e-306],
       [2.22522597e-306, 6.23053614e-307, 9.34608432e-307,
        2.56765117e-312]])

In [5]:
## 0부터 10개의 배열이 생성된다
np.arange(10)

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [7]:
# -5에서 5까지 1씩 증가하는 배열이 생성된다
np.arange(-5,5,1)

array([-5, -4, -3, -2, -1,  0,  1,  2,  3,  4])

## ndarray 배열의 모양, 차수, 데이터 타입을 확인하는 방법

In [11]:
arr = np.array([[1,2,3],[4,5,6]])

In [12]:
print(arr)

[[1 2 3]
 [4 5 6]]


In [13]:
arr.shape

(2, 3)

In [14]:
## arr의 배열은 몇차원 데이터인가?
arr.ndim

2

In [15]:
## arr의 배열은 어떤 타입의 데이터인가?
arr.dtype

dtype('int32')

## ndarray 배열의 타입 변환
astype 함수를 사용해서 정수형 데이터 배열을 실수형 데이터 배열로 변환하기

In [16]:
arr_int = np.array([1,2,3,4])

In [46]:
arr_int.dtype

dtype('int64')

In [18]:
arr_float = arr_int.astype(np.float64)

In [45]:
arr_float.dtype

dtype('float64')

In [21]:
print(arr_float)

[1. 2. 3. 4.]


### 문자형 데이터 배열을 정수형 데이터 배열로 변화하기

In [22]:
arr_str = np.array(['1','2','3'])

In [23]:
## 유니코드 타입으로 저장되어있다
arr_str.dtype

dtype('<U1')

In [24]:
arr_int = arr_str.astype(np.int64)

In [25]:
arr_int.dtype

dtype('int64')

![Built-in Python types, ndarray data types](images/numpy-1.png)

![Data Types](images/numpy-2.png)

## ndarray 배열의 연산
기본 연산자와 함수를 통해 배열 연산하기

In [26]:
arr1 = np.array([[1,2],[3,4]])
arr2 = np.array([[5,6],[7,8]])

In [27]:
arr1 + arr2

array([[ 6,  8],
       [10, 12]])

In [28]:
np.add(arr1, arr2)

array([[ 6,  8],
       [10, 12]])

In [29]:
arr1 * arr2

array([[ 5, 12],
       [21, 32]])

In [30]:
np.multiply(arr1, arr2)

array([[ 5, 12],
       [21, 32]])

### numpy 배열의 연산은 연산자(+, =, *, /)나 함수(add, substract, multiply, divide)가능합니다.

![Matrix Multiplication](images/numpy-6.png)
https://towardsdatascience.com/a-complete-beginners-guide-to-matrix-multiplication-for-data-science-with-python-numpy-9274ecfc1dc6

In [31]:
arr1.dot(arr2)

array([[19, 22],
       [43, 50]])

In [32]:
np.dot(arr1, arr2)

array([[19, 22],
       [43, 50]])

앞에서 살펴본 것과 같이 Numpy 배열에 기본 연산자 *를 통한 곱은 행력의 곱이 아닌 배열의 요소의 곱이다. 따라서 행렬의 곱을 수행하기 위해서는  dot  함수를 사용해야 한다.

## ndarray 배열 슬라이싱 하기

In [35]:
arr = np.array([[1,2,3],[4,5,6],[7,8,9]])

In [36]:
## 첫번째 배열에서 두번째 배열까지 보여주되 1에서 3까지만 보여달라
## 배열의 필요한 부분만 자르는 것을 '슬라이싱'이라고 한다
arr_1 = arr[:2,1:3]

In [37]:
print(arr_1)

[[2 3]
 [5 6]]


In [38]:
arr_1[0,0] = 5

In [39]:
arr_1

array([[5, 3],
       [5, 6]])

In [40]:
## 첫번쨰 배열의 1번째 속성값도 수정된 것을 확인할 수 있다
arr

array([[1, 5, 3],
       [4, 5, 6],
       [7, 8, 9]])

In [41]:
arr = np.array([[1,2,3],[4,5,6]])

In [42]:
idx = arr > 3

In [43]:
## 배열에 조건을 주면 bool 타입으로 보여준다
print(idx)

[[False False False]
 [ True  True  True]]


In [44]:
## 조건에 맞는 값만 보여준다
print(arr[idx])

[4 5 6]


idx의 값들은 arr의 값이 3보다 큰 것인지(True), 같거나 작은 것인지(False) 나타낸다.



### Wine Quality 데이터 설명
Wine Quality 데이터에는 포르투갈의 레드 와인과 화이트 와인 두 개의 데이터 세시이 있다. 각각의 데이터 예시 안에는 물리화학적 실험에 대한 12개의 변수들이 있다. 변수에 대한 설명은 다음과 같다.

![Wine Quality 데이터 변수 설명](images/numpy-3.png)

In [48]:
import numpy as np

redwine = np.loadtxt(fname='samples/winequality-red.csv', delimiter=';', skiprows=1)

In [49]:
print(redwine)

[[ 7.4    0.7    0.    ...  0.56   9.4    5.   ]
 [ 7.8    0.88   0.    ...  0.68   9.8    5.   ]
 [ 7.8    0.76   0.04  ...  0.65   9.8    5.   ]
 ...
 [ 6.3    0.51   0.13  ...  0.75  11.     6.   ]
 [ 5.9    0.645  0.12  ...  0.71  10.2    5.   ]
 [ 6.     0.31   0.47  ...  0.66  11.     6.   ]]


In [51]:
redwine.shape

(1599, 12)

### Wine Quality 데이터를 활용한 기초 통계 분석
NumPy에는 기본적인 배열 통계 메서드가 있다. sum, mean과 같은 기본적인 통계 메서드를 통해 3.2에서 불러온 redwine 데이터의 기초적인 통계 분석을 해본다. 실습할 통계 메서드는 다음과 같다.

![배열 통계 메서드](images/numpy-4.png)

In [52]:
print(redwine.sum())

152084.78194


In [53]:
print(redwine.mean())

7.926036165311652


In [57]:
# axis = 좌표의 축
print(redwine.mean(axis=0))

[ 8.31963727  0.52782051  0.27097561  2.5388055   0.08746654 15.87492183
 46.46779237  0.99674668  3.3111132   0.65814884 10.42298311  5.63602251]


In [56]:
# redwine의 전체 데이터에서 0번째 항목의 평균을 계산해라
redwine[:, 0].mean()

8.31963727329581

redwine.sum(): redwine 데이터 셋에 있는 모든 데이터의 합 <br>
redwine.mean(): redwine 데이터 셋에 있는 모든 데이터의 평균 <br>
redwine.mean(axis=0): redwine 데이터 셋의 열별(axis=0은 열을 뜻한다.) 평균, 이를 통해 각 변수별 평균을 알 수 있다. Fixed acidity의 평균은 8.31963722 이며, volatile acidity의 평균은 0.52782051 이다. <br>
redwine[:0.mean(): redwine 데이터 셋의 첫번째 열 (0번째 열)의 모든 값들의 평균을 뜻한다. 즉, fixed acidity의 평균만 알 수 있다.

In [58]:
# 0번 축을 기준으로 각 컬럼의 최대값을 구하라
print(redwine.max(axis=0))

[ 15.9       1.58      1.       15.5       0.611    72.      289.
   1.00369   4.01      2.       14.9       8.     ]


In [59]:
print(redwine.min(axis=0))

[4.6     0.12    0.      0.9     0.012   1.      6.      0.99007 2.74
 0.33    8.4     3.     ]
